util/app/path/path.go

```diff
@@ -27,3 +28,56 @@ func Path(root, path string) (string, error) {
 	}
 	return appPath, nil
 }
+
+const outOfBoundsError string = "Out of bounds symlink found"
+
+// ContainsOutOfBoundsSymlinks determines if basePath contains any symlinks that
+// are absolute or point to a path outside of the basePath.
+func ContainsOutOfBoundsSymlinks(basePath string) (bool, string, error) {
+	found := ""
+	var err error
+	absBasePath, err := filepath.Abs(basePath)
+	if err != nil {
+		return false, "", err
+	}
+	err = filepath.Walk(absBasePath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.Mode()&os.ModeSymlink == os.ModeSymlink {
+			// We don't use filepath.EvalSymlinks because it fails without returning a path
+			// if the target doesn't exist.
+			target, err := os.Readlink(path)
+			if err != nil {
+				return err
+			}
+			repoPath, err := filepath.Rel(absBasePath, path)
+			if err != nil {
+				return err
+			}
+			if filepath.IsAbs(target) {
+				found = repoPath
+				return errors.New(outOfBoundsError)
+			}
+			dir := filepath.Dir(path)
+			joined := filepath.Join(dir, target)
+			rel, err := filepath.Rel(absBasePath, joined)
+			if err != nil {
+				return err
+			}
+			if strings.Contains(rel, "..") {
```

Comment:
This is tricky, because `a..b` is a valid filename.

What we've done in other places is calculate the absolute path and then making sure its prefix is the expected root path (with the trailing `/`). https://github.com/argoproj/argo-cd/blob/master/util/io/files/util.go#L72-L90

Reply:
Ah crap, you're right! I forgot about that. Gonna have to think about this one. Good catch!

Reply:
I wonder if `startswith(.. + path separator)` would be sufficient.